torchmultimodal/modules/encoders/swin_transformer_3d_encoder.py

```diff
@@ -0,0 +1,567 @@
+    It will use common PatchEmbed3d for image and video,
+    for single view depth image it will have separate embedding for the depth channel
+    and add the embedding result with the RGB channel
+    reference: https://arxiv.org/abs/2201.08377
+
+    Args:
+        patch_size (Tuple[int, int, int]): Patch token size. Default: (2, 4, 4)
+        embed_dim (int): Number of linear projection output channels. Default: 96
+        norm_layer (nn.Module, optional): Normalization layer. Default: None
+    """
+
+    def __init__(
+        self,
+        patch_size: Tuple[int, int, int] = (2, 4, 4),
+        embed_dim: int = 96,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+    ):
+        super().__init__()
+        self.patch_embed = PatchEmbed3d(
+            patch_size=patch_size,
+            embed_dim=embed_dim,
+            norm_layer=norm_layer,
+        )
+
+        self.depth_patch_embed = PatchEmbed3d(
+            patch_size=patch_size,
+            in_chans=1,
+            embed_dim=embed_dim,
+            norm_layer=norm_layer,
+        )
+
+    def forward(self, x):
+        # x: B C D H W
+        # Note: D here represent time
+        assert x.ndim == 5
+        has_depth = x.shape[1] == 4
+
+        if has_depth:
+            x_rgb = self.patch_embed(x[:, :3, ...])
+            x_d = self.depth_patch_embed(x[:, 3:, ...])
+            x = x_rgb + x_d
+        else:
+            x = self.patch_embed(x)
+        return x
+
+
+class SwinTransformer3dEncoder(nn.Module):
+    """
+    Implements 3D Swin Transformer from the `"Video Swin Transformer" <https://arxiv.org/abs/2106.13230>`_ paper.
+    Args:
+        patch_size (Tuple[int, int, int]): Patch size.
+        embed_dim (int): Patch embedding dimension.
+        depths (List(int)): Depth of each Swin Transformer layer.
+        num_heads (List(int)): Number of attention heads in different layers.
+        window_size (Tuple[int, int, int]): Window size. Default: (2, 7, 7).
+        mlp_ratio (float): Ratio of mlp hidden dim to embedding dim. Default: 4.0.
+        dropout (float): Dropout rate. Default: 0.0.
+        attention_dropout (float): Attention dropout rate. Default: 0.0.
+        stochastic_depth_prob (float): Stochastic depth rate. Default: 0.0.
+        num_classes (int): Number of classes for classification head. Default: 1000.
+        block (nn.Module, optional): SwinTransformer Block. Default: None.
+        norm_layer (nn.Module, optional): Normalization layer. Default: None.
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        depths: List[int],
+        num_heads: List[int],
+        window_size: Tuple[int, int, int] = (2, 7, 7),
+        mlp_ratio: float = 4.0,
+        dropout: float = 0.0,
+        attention_dropout: float = 0.0,
+        stochastic_depth_prob: float = 0.0,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+        block: Optional[Callable[..., nn.Module]] = None,
```

Comment:
Could you elaborate more or maybe give some examples on this? What do you mean by separate function which takes in flat params?

Reply:
You can check our newly uploaded wiki about developer guide and let us know if it makes sense (ping you separately in chat). But for convenience here is the relevant block:

>If you have class A which uses instances of class B and C, the init should take in instantiations of B and C (instead of params for B and C and instantiating within init). This will give users flexibility to pass in their own instantiations and now A is independent of any init signature changes of B and C.

Reply:
I think this method of passing instantiated object would be difficult for a dynamically instantiated layers.

For example here SwinTransformer3dEncoder have a stack of SwinTransformer3dBlock where the size of the stack depend on the config and the parameter of each SwinTransformer3dBlock depend on the previous layer. Any suggestions for this?

Reply:
>For example here SwinTransformer3dEncoder have a stack of SwinTransformer3dBlock where the size of the stack depend on the config 

Are you referring to `depth`? Since it defines the structure of the SwinTransformer3dEncoder and is not related to the classes it relies on,  I think it's ok to leave it as an argument of the constructor.

> and the parameter of each SwinTransformer3dBlock depend on the previous layer.

What if someone wants to use a variant of SwinTransformer3dBlock? Would it be possible to allow the user to pass the class name instead of the Encoder assigns it? Also, why not making SwinTransformer3dBlock and `partial(nn.LayerNorm, eps=1e-5)` the default values for `block` and `norm_layer`?

I agree the approach described by our contributing guide is not applicable here since class B and C (the building blocks) are dynamically instantiated inside class A (the class that uses them). But ultimately, the idea is to allow customization and reusability. 
If you think SwinTransformer3dEncoder can only use SwinTransformer3dBlock but nothing else then it makes less sense to allow that flexbility.

Reply:
The same consideration applies to `SwinTransformerBlock3d` which also takes in a few lower level classes